From 8e74f9a50a6fa466021ba53a5d1500fd33d6b80a Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Tue, 2 Dec 2025 16:02:51 +0800
Subject: [PATCH 024/105] mtk: mt76: mt7996: rework chainmask

It is more convenient for the driver to leave the chainmask unshifted, as
the shifted chainmask is only used for radio->antenna_mask and
wiphy->available_antennas_t(r)x.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>

tx_ant & rx_ant are the accumulated antenna masks for each radio.
Therefore, after refactoring, since phy->mt76->chainmask is unshifted,
we should use radio->antenna_mask (the shifted version of chainmask) to
parse the mask for each radio.
For example,
tx_ant = 0x173 => radio 0: 0x3, radio 1: 0x7, radio 2: 0x1
phy->chainmask = 0xf for each band
radio->antenna_mask: radio 0: 0xf, radio 1: 0xf0, radio 2: 0xf00
The modified chainmask for radio 1 should be (0x173 & 0xf0) >> 4 = 0x7

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>

After reworking, chainmask becomes unshifted.
Therefore, the chainmask should be shifted to retrieve the combined
antenna mask for each radio.
To get chainshift in the common part, the chainshift array should be
moved from mt7996_dev to mt76_dev.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mac80211.c      |  8 +++++---
 mt76.h          |  2 ++
 mt7996/eeprom.c |  7 +++----
 mt7996/init.c   |  5 +++--
 mt7996/main.c   | 20 ++++++++++++--------
 mt7996/mcu.c    |  2 +-
 mt7996/mt7996.h |  2 --
 7 files changed, 26 insertions(+), 20 deletions(-)

diff --git a/mac80211.c b/mac80211.c
index 9d6f89e7a..50d14275f 100644
--- a/mac80211.c
+++ b/mac80211.c
@@ -1957,9 +1957,11 @@ int mt76_get_antenna(struct ieee80211_hw *hw, int radio_idx, u32 *tx_ant,
 
 	mutex_lock(&dev->mutex);
 	*tx_ant = 0;
-	for (i = 0; i < ARRAY_SIZE(dev->phys); i++)
-		if (dev->phys[i] && dev->phys[i]->hw == hw)
-			*tx_ant |= dev->phys[i]->chainmask;
+	for (i = 0; i < ARRAY_SIZE(dev->phys); i++) {
+		phy = dev->phys[i];
+		if (phy && phy->hw == hw)
+			*tx_ant |= phy->chainmask << dev->chainshift[phy->band_idx];
+	}
 	*rx_ant = *tx_ant;
 	mutex_unlock(&dev->mutex);
 
diff --git a/mt76.h b/mt76.h
index 646cc8e57..44c9c6192 100644
--- a/mt76.h
+++ b/mt76.h
@@ -1039,6 +1039,8 @@ struct mt76_dev {
 	};
 
 	atomic_t bus_hung;
+
+	u8 chainshift[__MT_MAX_BAND];
 };
 
 /* per-phy stats.  */
diff --git a/mt7996/eeprom.c b/mt7996/eeprom.c
index 8f6077291..681aa9d0f 100644
--- a/mt7996/eeprom.c
+++ b/mt7996/eeprom.c
@@ -311,12 +311,11 @@ int mt7996_eeprom_parse_hw_cap(struct mt7996_dev *dev, struct mt7996_phy *phy)
 
 	mphy->antenna_mask = BIT(nss) - 1;
 	phy->orig_antenna_mask = mphy->antenna_mask;
-	mphy->chainmask = (BIT(path) - 1) << dev->chainshift[band_idx];
+	mphy->chainmask = BIT(path) - 1;
 	phy->orig_chainmask = mphy->chainmask;
-	dev->chainmask |= mphy->chainmask;
 	if (band_idx < MT_BAND2)
-		dev->chainshift[band_idx + 1] = dev->chainshift[band_idx] +
-						hweight16(mphy->chainmask);
+		dev->mt76.chainshift[band_idx + 1] = dev->mt76.chainshift[band_idx] +
+						     hweight16(mphy->chainmask);
 
 	return mt7996_eeprom_parse_band_config(phy);
 }
diff --git a/mt7996/init.c b/mt7996/init.c
index 0fcaaaa40..3215e4757 100644
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -463,10 +463,11 @@ mt7996_init_wiphy_band(struct ieee80211_hw *hw, struct mt7996_phy *phy)
 	radio->n_freq_range = 1;
 	radio->iface_combinations = &if_comb;
 	radio->n_iface_combinations = 1;
+	radio->antenna_mask = phy->mt76->chainmask << dev->mt76.chainshift[phy->mt76->band_idx];
 	hw->wiphy->n_radio++;
 
-	wiphy->available_antennas_rx |= phy->mt76->chainmask;
-	wiphy->available_antennas_tx |= phy->mt76->chainmask;
+	wiphy->available_antennas_rx |= radio->antenna_mask;
+	wiphy->available_antennas_tx |= radio->antenna_mask;
 
 	mt76_set_stream_caps(phy->mt76, true);
 	mt7996_set_stream_vht_txbf_caps(phy);
diff --git a/mt7996/main.c b/mt7996/main.c
index 815a4ad42..b152f1db4 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -1667,28 +1667,32 @@ mt7996_set_antenna(struct ieee80211_hw *hw, int radio_idx,
 		   u32 tx_ant, u32 rx_ant)
 {
 	struct mt7996_dev *dev = mt7996_hw_dev(hw);
+	struct mt7996_phy *phy;
+	struct wiphy_radio *radio;
 	int i;
 
 	if (tx_ant != rx_ant)
 		return -EINVAL;
 
 	for (i = 0; i < hw->wiphy->n_radio; i++) {
-		struct mt7996_phy *phy = dev->radio_phy[i];
+		phy = dev->radio_phy[i];
+		radio = &dev->radios[i];
 
-		if (!(tx_ant & phy->orig_chainmask))
+		if (!(tx_ant & radio->antenna_mask))
 			return -EINVAL;
 	}
 
 	mutex_lock(&dev->mt76.mutex);
 
 	for (i = 0; i < hw->wiphy->n_radio; i++) {
-		struct mt7996_phy *phy = dev->radio_phy[i];
-		u8 band_idx = phy->mt76->band_idx;
-		u8 shift = dev->chainshift[band_idx];
+		int shift;
+
+		phy = dev->radio_phy[i];
+		radio = &dev->radios[i];
+		shift = dev->mt76.chainshift[phy->mt76->band_idx];
 
-		phy->mt76->chainmask = tx_ant & phy->orig_chainmask;
-		phy->mt76->antenna_mask = (phy->mt76->chainmask >> shift) &
-					  phy->orig_antenna_mask;
+		phy->mt76->chainmask = (tx_ant & radio->antenna_mask) >> shift;
+		phy->mt76->antenna_mask = phy->mt76->chainmask & phy->orig_antenna_mask;
 
 		mt76_set_stream_caps(phy->mt76, true);
 		mt7996_set_stream_vht_txbf_caps(phy);
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index 97b8d6dfd..b154353d8 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -3863,7 +3863,7 @@ int mt7996_mcu_set_chan_info(struct mt7996_phy *phy, u16 tag)
 		.center_ch = ieee80211_frequency_to_channel(freq1),
 		.bw = mt76_connac_chan_bw(chandef),
 		.tx_path_num = hweight16(phy->mt76->chainmask),
-		.rx_path = mt7996_rx_chainmask(phy) >> dev->chainshift[band_idx],
+		.rx_path = mt7996_rx_chainmask(phy),
 		.band_idx = band_idx,
 		.channel_band = ch_band[chandef->chan->band],
 	};
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
index f850be874..89bac2d5e 100644
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -406,8 +406,6 @@ struct mt7996_dev {
 	struct cfg80211_chan_def rdd2_chandef;
 	struct mt7996_phy *rdd2_phy;
 
-	u16 chainmask;
-	u8 chainshift[__MT_MAX_BAND];
 	u32 hif_idx;
 
 	struct work_struct init_work;
-- 
2.45.2

